### PR TITLE
TorchMLIR eager mode with IREE backend

### PR DESCRIPTION
**Pull request by makslevental**
Eager mode hidden behind `SharkMode` (see `shark/examples/eager_mode.py`).

```
TorchMLIRTensor([[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]], backend=EagerModeRefBackend)
TorchMLIRTensor([[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]], backend=EagerModeRefBackend)
```

Note also that you need https://github.com/makslevental/torch-mlir/tree/add_device as your activated torch-mlir version.

### CODE REVIEW

File: shark/examples/eager_mode.py
Comment (nirvedhmeshram) [inline shark/examples/eager_mode.py:123]: A lot of these functions are never called?

File: shark/iree_eager_backend.py
Comment (nirvedhmeshram) [inline shark/iree_eager_backend.py:19]: I dont see this function used..

File: shark/examples/eager_mode.py
Comment (nirvedhmeshram) [inline shark/examples/eager_mode.py:1]: do we want any copyright/licensing informatiom here and in all the other new files? Probably a question for @powderluv

File: shark/iree_eager_backend.py
Comment (nirvedhmeshram) [inline shark/iree_eager_backend.py:33]: A comment describing this class would be helpful for understanding the code

File: shark/iree_eager_backend.py
Comment (nirvedhmeshram) [inline shark/iree_eager_backend.py:34]: A comment describing this class would be helpful for understanding the code

File: shark/examples/eager_mode.py
Comment (makslevental) [reply shark/examples/eager_mode.py:123]: they're all for exercising the iree backend, which is blocked

File: shark/iree_eager_backend.py
Comment (makslevental) [reply shark/iree_eager_backend.py:19]: good catch - yea that's from before my refactor of the `add_device` torch-mlir pr.

File: shark/examples/eager_mode.py
Comment (nirvedhmeshram) [reply shark/examples/eager_mode.py:123]: I see, is the intention to land this once they are unblocked?

Comment (powderluv) [reply shark/examples/eager_mode.py:123]: Latest shark build should have a fix

**Review by nirvedhmeshram — APPROVED** [review]
LGTM!

**Comment by makslevental** [comment]
@nirvedhmeshram i don't have commit access so feel free to merge (or to add me)

File: shark/examples/eager_mode.py
Comment (nirvedhmeshram) [reply shark/examples/eager_mode.py]: actually, if IREE is unblocked now should this be uncommented?

Comment (nirvedhmeshram) [reply shark/examples/eager_mode.py]: nevermind was looking at old stuff

**Comment by nirvedhmeshram** [comment]
> @nirvedhmeshram i don't have commit access so feel free to merge (or to add me)

@makslevental  Can you please look into the CI failure? Looks like something to do with eager mode.

**Comment by makslevental** [comment]
> > @nirvedhmeshram i don't have commit access so feel free to merge (or to add me)
> 
> @makslevental Can you please look into the CI failure? Looks like something to do with eager mode.

It's because this is against a branch/PR that hasn't been merged into torch-mlir (none of the imported torch_mlir objects exist in the current release of torch-mlir). So I guess we hold off on merging until upstream gets merged in.

**Comment by powderluv** [comment]
Upstream pytorch ? It should be running on the latest nightly. Or is there a more recent commit we need from upstream ?

File: shark/shark_runner.py
Comment (powderluv) [inline shark/shark_runner.py:164]: Is this the best name ? What is non-SharkMode ? Stock IREE  ?

**Review by powderluv — APPROVED** [review]
minor nit

File: shark/shark_runner.py
Comment (makslevental) [reply shark/shark_runner.py:164]: oh well i have no idea - i just thought of shark as an alternate to "torch-mlir mode", which is just tracing+compilation+refbackend. I can rename to whatever you think is best.

Comment (powderluv) [reply shark/shark_runner.py:164]: In the context of this maybe SharkEagerMode ?

Comment (makslevental) [reply shark/shark_runner.py:164]: how about `JawsMode` since hungry <=> eager?

i'm only half joking...

Comment (powderluv) [reply shark/shark_runner.py:164]: We should use JAWS for something

**Pull request merged**
